src/datasets/io/json.py

```diff
@@ -84,11 +90,17 @@ def write(self) -> int:
         lines = self.to_json_kwargs.pop("lines", True)
         compression = self.to_json_kwargs.pop("compression", None)
 
-        if compression not in [None, "gzip"]:
-            raise NotImplementedError(f"Datasets currently does not support {compression} compression")
+        if compression not in [None, "infer", "gzip", "bz2", "xz", "zip"]:
+            raise NotImplementedError(f"`datasets` currently does not support {compression} compression")
+
+        # Infer compression from the filename/URL extension
+        if compression == "infer":
+            for compression_format, extension in self._compression_to_extension.items():
+                if self.path_or_buf.lower().endswith(extension):
+                    compression = compression_format
 
         if isinstance(self.path_or_buf, (str, bytes, os.PathLike)):
-            with IOHandler(self.path_or_buf, "wb", compression) as buffer:
+            with FileWriteHandler(self.path_or_buf, "wb", compression) as buffer:
```

Comment:
I just changed this to:
`with fsspec.open(self.path_or_buf, "wb", compression=compression,) as buffer:`

Reply:
Yeah, for now we can go with `gzip`, `bz2`, `xz`, and `infer`. `compresslevel` also let's handle in next PR (as of now I see same speed for all values)

Reply:
Ok ! Would it be possible to add a test before we merge this one ?

Reply:
Create another compressed file and check file sizes?